In [1]:
import pandas as pd
import numpy as np

# COMPLETEZZA

## COMPLETEZZA POKEMON

Più che considerare la completezza a livello di intera tabella, credo sia sensato valutare colonna per colonna, dal momento che per alcuni attributi la presenza di un NaN indica che quel dato non esiste per quella specifica entità, e non che è un dato mancante (per esempio, Bulbasaur ha un NaN nella colonna 'Evolves_from' solo perchè non evolve da nessuno, non perchè manchi quella informazione).

In [2]:
pokemon = pd.read_csv('Dataset integrati 2.0/df_pokemon.csv')
pokemon

#                  Name    Species       Variant  Generation  \
0       1             Bulbasaur  Bulbasaur           NaN           1   
1       2               Ivysaur    Ivysaur           NaN           1   
2       3              Venusaur   Venusaur           NaN           1   
3       3         Mega Venusaur   Venusaur          Mega           6   
4       3   Venusaur Gigantamax   Venusaur    Gigantamax           8   
...   ...                   ...        ...           ...         ...   
1093  896             Glastrier  Glastrier           NaN           8   
1094  897             Spectrier  Spectrier           NaN           8   
1095  898               Calyrex    Calyrex           NaN           8   
1096  898  Calyrex Shadow Rider    Calyrex  Shadow Rider           8   
1097  898     Calyrex Ice Rider    Calyrex     Ice Rider           8   

         Rarity Evolves_from Has_gender_diff    Type1   Type2  ...  Attack  \
0        Common          NaN           False    Grass  Poison  ...    49.0   
1        Common    Bulbasaur           False    Grass  Poison  ...    62.0   
2        Common      Ivysaur            True    Grass  Poison  ...    82.0   
3        Common     Venusaur            True    Grass  Poison  ...   100.0   
4           NaN     Venusaur             NaN    Grass  Poison  ...     NaN   
...         ...          ...             ...      ...     ...  ...     ...   
1093  Legendary          NaN           False      Ice     NaN  ...   145.0   
1094  Legendary          NaN           False    Ghost     NaN  ...    65.0   
1095  Legendary          NaN           False  Psychic   Grass  ...    80.0   
1096  Legendary          NaN           False  Psychic   Ghost  ...    85.0   
1097  Legendary          NaN           False  Psychic     Ice  ...   165.0   

      Defense  Sp. Atk  Sp. Def  Speed  \
0        49.0     65.0     65.0   45.0   
1        63.0     80.0     80.0   60.0   
2        83.0    100.0    100.0   80.0   
3       123.0    122.0    120.0   80.0   
4         NaN      NaN      NaN    NaN   
...       ...      ...      ...    ...   
1093    130.0     65.0    110.0   30.0   
1094     60.0    145.0     80.0  130.0   
1095     80.0     80.0     80.0   80.0   
1096     80.0    165.0    100.0  150.0   
1097    150.0     85.0    130.0   50.0   

                                              image_url  \
0     https://img.pokemondb.net/sprites/sword-shield...   
1     https://img.pokemondb.net/sprites/sword-shield...   
2     https://img.pokemondb.net/sprites/sword-shield...   
3     https://img.pokemondb.net/sprites/sword-shield...   
4     https://archives.bulbagarden.net/media/upload/...   
...                                                 ...   
1093  https://img.pokemondb.net/sprites/sword-shield...   
1094  https://img.pokemondb.net/sprites/sword-shield...   
1095  https://img.pokemondb.net/sprites/sword-shield...   
1096  https://img.pokemondb.net/sprites/sword-shield...   
1097  https://img.pokemondb.net/sprites/sword-shield...   

                                          VGC2022_rules Monthly Usage (k)  \
0                                             Permitted               NaN   
1                                             Permitted               NaN   
2                                             Permitted               204   
3                                                Banned               NaN   
4                                    Gigantamax Allowed               NaN   
...                                                 ...               ...   
1093                                          Permitted                13   
1094                                          Permitted                19   
1095  Restricted (players can include two Restricted...               NaN   
1096                                                NaN               363   
1097                                                NaN               294   

     Usage Percent (%) Monthly Rank  
0                  NaN       

In [3]:
pkmn_column_nan = pd.DataFrame(pokemon.isna().sum(), columns=['Total NaN'])
pkmn_column_nan

Total NaN
#                          0
Name                       0
Species                    0
Variant                  861
Generation                 0
Rarity                    32
Evolves_from             543
Has_gender_diff           32
Type1                      0
Type2                    510
Total                     32
HP                        32
Attack                    32
Defense                   32
Sp. Atk                   32
Sp. Def                   32
Speed                     32
image_url                  0
VGC2022_rules            381
Monthly Usage (k)        756
Usage Percent (%)        756
Monthly Rank             756

In [4]:
#Calcolo la percentuale di NaN per singola colonna, considerando che abbiamo 1098 righe
pkmn_column_nan['% NaN'] = (pkmn_column_nan['Total NaN']/1098)*100
pkmn_column_nan = pkmn_column_nan.round(decimals=2)
pkmn_column_nan

Total NaN  % NaN
#                          0   0.00
Name                       0   0.00
Species                    0   0.00
Variant                  861  78.42
Generation                 0   0.00
Rarity                    32   2.91
Evolves_from             543  49.45
Has_gender_diff           32   2.91
Type1                      0   0.00
Type2                    510  46.45
Total                     32   2.91
HP                        32   2.91
Attack                    32   2.91
Defense                   32   2.91
Sp. Atk                   32   2.91
Sp. Def                   32   2.91
Speed                     32   2.91
image_url                  0   0.00
VGC2022_rules            381  34.70
Monthly Usage (k)        756  68.85
Usage Percent (%)        756  68.85
Monthly Rank             756  68.85

Sulle colonne relative alle statistiche, oltre a 'Rarity' e 'Has_gender_diff' abbiamo un 2.91% di NaN, che fanno però riferimento alle 32 varianti Gigantamax presenti nel dataset. Ci mancano quindi le informazioni di queste specifiche varianti.

Per quanto riguarda le colonne 'Variant', 'Evolves_from' e 'Type2' l'elevata percentuale di NaN non indica una mancanza di informazione, quanto più il fatto che quasi l'80% dei Pokemon non hanno varianti, circa il 50% sono Pokemon base che non evolvono da nessun altro Pokemon e poco meno del 47% dei Pokemon non hanno un secondo tipo (infatti la colonna 'Type1' non presenta NaN).

Non abbiamo informazioni per l'attributo 'VGC2022_rules' per quasi il 35% dei Pokemon. C'è poi il 68.85% di NaN nelle colonne 'Monthly Usage (k)', 'Usage Percent (%)' e 'Monthly Rank': questo dato, però, non è particolarmente preoccupante, poichè i Pokemon che per la colonna 'VGC2022_rules' hanno valore 'Banned' o un NaN non possono avere informazioni per queste tre colonne. Allo stesso modo, la presenza di NaN per gli altri Pokemon denota semplicemente come tali Pokemon non sia utilizzati in competitivo.

In [50]:
## Completezza intero dataset
print(pokemon.size)
print(pokemon.isna().sum().sum())
print(pokemon.isna().sum().sum() / pokemon.size * 100)

24156
4851
20.081967213114755


## COMPLETEZZA ITEMS

In [5]:
items = pd.read_csv('Dataset integrati 2.0/df_items.csv')
items

Name                                      Effect (alt.)  \
0    Ability Capsule  Switches a Pokémon between its two possible (n...   
1      Ability Patch                                                NaN   
2       Ability Urge   Forcibly activates a friendly Pokémon's ability.   
3          Abomasite  Held: Allows Abomasnow to Mega Evolve into Meg...   
4           Absolite  Held: Allows Absol to Mega Evolve into Mega Ab...   
..               ...                                                ...   
835     Yellow Scarf  Raises the holder's Toughness while in a contest.   
836     Yellow Shard       No effect. Can be traded for items or moves.   
837        Zap Plate  Held: Electric-Type moves from holder do 20% m...   
838             Zinc              Raises Special Defense and happiness.   
839        Zoom Lens  Held: Provides a 1/5 (20%) boost in accuracy i...   

                                                Effect  \
0    A capsule that allows a Pokémon with two Abili...   
1    Changes a Pokémon's ability to its Hidden Abil...   
2    When used, it activates the Ability of an ally...   
3      Enables Abomasnow to Mega Evolve during battle.   
4          Enables Absol to Mega Evolve during battle.   
..                                                 ...   
835         Raises holder's Tough aspect in a Contest.   
836  A small yellow shard. It appears to be from so...   
837  Increases power of Electric-type moves. Change...   
838               Increases Special Defense EVs by 10.   
839  Raises a move's accuracy if the holder moves a...   

                                      Usage Attributes Specific Category  \
0                                                  NaN          Vitamins   
1                                                  NaN          Vitamins   
2                                                  NaN   Miracle Shooter   
3                                                  NaN       Mega Stones   
4                                                  NaN       Mega Stones   
..                                                 ...               ...   
835                                ['holdable-active']           Scarves   
836                                    ['underground']      Collectibles   
837     ['holdable', 'holdable-active', 'underground']            Plates   
838  ['countable', 'consumable', 'usable-overworld'...          Vitamins   
839                    ['holdable', 'holdable-active']        Held Items   

    General Category                                          image_url  
0         Hold items  https://img.pokemondb.net/sprites/items/abilit...  
1      General items                    https://img.pokemondb.net/s.png  
2       Battle items  https://img.pokemondb.net/sprites/items/abilit...  
3         Hold items  https://img.pokemondb.net/sprites/items/abomas...  
4         Hold items  https://img.pokemondb.net/sprites/items/absoli...  
..               ...                                                ...  
835    General items  https://img.pokemondb.net/sprites/items/yellow...  
836    General items  https://img.pokemondb.net/sprites/items/yellow...  
837       Hold items  https://img.pokemondb.net/sprites/items/zap-pl...  
838         Medicine   https://img.pokemondb.net/sprites/items/zinc.png  
839       Hold items  https://img.pokemondb.net/sprites/items/zoom-l...  

[840 rows x 7 columns]

In [6]:
items_column_nan = pd.DataFrame(items.isna().sum(), columns=['Total NaN'])
items_column_nan

Total NaN
Name                       0
Effect (alt.)            224
Effect                    16
Usage Attributes         599
Specific Category          0
General Category           0
image_url                  0

In [7]:
#Calcolo la percentuale di NaN per singola colonna, considerando che abbiamo 840 righe
items_column_nan['% NaN'] = (items_column_nan['Total NaN']/840)*100
items_column_nan = items_column_nan.round(decimals=2)
items_column_nan

Total NaN  % NaN
Name                       0   0.00
Effect (alt.)            224  26.67
Effect                    16   1.90
Usage Attributes         599  71.31
Specific Category          0   0.00
General Category           0   0.00
image_url                  0   0.00

Per 16 entità del dataset Items manca l'effetto. Per poco più del 26% degli Items manca l'effetto alternativo.  La colonna 'Usage Attributes' presenta invece il 71.31% di valori mancanti: era un attributo sparso sin dall'inizio..

In [52]:
## Completezza intero dataset
print(items.size)
print(items.isna().sum().sum())
print(items.isna().sum().sum() / items.size * 100)

5880
839
14.268707482993198


## COMPLETEZZA MOVES

In [8]:
moves = pd.read_csv('Dataset integrati 2.0/df_moves.csv')
moves

Name  Introducted_in    Type  Power Acc.    PP Damage_class  \
0              Absorb               1   Grass   20.0  100  25.0      Special   
1          Accelerock               7    Rock   40.0  100  20.0     Physical   
2                Acid               1  Poison   40.0  100  30.0      Special   
3          Acid Armor               1  Poison    NaN  NaN  20.0       Status   
4          Acid Spray               5  Poison   40.0  100  20.0      Special   
..                ...             ...     ...    ...  ...   ...          ...   
816      G-Max Finale               8   Fairy    NaN    ∞   5.0          NaN   
817  G-Max Steelsurge               8   Steel    NaN    ∞   5.0          NaN   
818   G-Max Depletion               8  Dragon    NaN    ∞   5.0          NaN   
819    G-Max One Blow               8    Dark    NaN    ∞   5.0          NaN   
820  G-Max Rapid Flow               8   Water    NaN    ∞   5.0          NaN   

                                                Effect  Prob. (%)  
0     User recovers half the HP inflicted on opponent.        NaN  
1                                  User attacks first.        NaN  
2                May lower opponent's Special Defense.       10.0  
3                       Sharply raises user's Defense.        NaN  
4           Sharply lowers opponent's Special Defense.      100.0  
..                                                 ...        ...  
816  Heals the user and its allies by 1/6 their max...        NaN  
817  Scatters sharp spikes around the field Works l...        NaN  
818  Takes away 2 PP from the last move the target ...        NaN  
819  Hits the target even if it is protected by a p...        NaN  
820  Hits the target even if it is protected by a p...        NaN  

[821 rows x 9 columns]

In [9]:
moves_column_nan = pd.DataFrame(moves.isna().sum(), columns=['Total NaN'])
moves_column_nan

Total NaN
Name                    0
Introducted_in          0
Type                    0
Power                 346
Acc.                  216
PP                     20
Damage_class           33
Effect                 31
Prob. (%)             591

In [10]:
#Calcolo la percentuale di NaN per singola colonna, considerando che abbiamo 821 righe
moves_column_nan['% NaN'] = (moves_column_nan['Total NaN']/821)*100
moves_column_nan = moves_column_nan.round(decimals=2)
moves_column_nan

Total NaN  % NaN
Name                    0   0.00
Introducted_in          0   0.00
Type                    0   0.00
Power                 346  42.14
Acc.                  216  26.31
PP                     20   2.44
Damage_class           33   4.02
Effect                 31   3.78
Prob. (%)             591  71.99

Per circa il 4% delle mosse non abbiamo informazioni sull'attributo 'Effect' e per un altro 4% non abbiamo informazione riguardo l'attributo 'Damage_class': nel primo caso è plausibile che siano mosse senza un effetto secondario (la mossa 'Aqua Tail', per esempio, è una mossa che infligge solamente danno, senza applicare alcun effetto aggiuntivo (https://bulbapedia.bulbagarden.net/wiki/Aqua_Tail_(move))), nel secondo caso invece si tratta delle mosse G-Max che non hanno una categoria specifica di danno poichè la mossa può essere sia fisica che speciale a seconda della mossa da cui parte (https://bulbapedia.bulbagarden.net/wiki/Max_Move).

Il 72% delle mosse ha valore mancante per l'attributo 'Prob. (%)': questo però dipende dal fatto che solo alcune mosse hanno degli effetti (come può essere uno status o un drop di una statistica) che si attivano con una certa probabilità, mentre gli altri effetti sono peculiari della mossa e si attivano sempre semplicemente eseguendola.

Per quanto riguarda 'Power' la presenza dei NaN è dettata dalla presenza delle mosse status che non infliggono danni, mentre per la colonna 'Acc.' la presenza dei NaN è conseguenza del fatto che alcune mosse non possono fallire (le mosse per cui 'Acc.'=100 sono mosse che, di base, non falliscono mai, ma in presenza di un drop della precisione di un Pokemon in battaglia possono effettivamente fallire, mentre quelle con valore NaN non vengono influenzate da eventuali drop e non falliscono mai).

Abbiamo poi informazioni mancanti riguardo i PP di 20 mosse.

In [53]:
## Completezza intero dataset
print(moves.size)
print(moves.isna().sum().sum())
print(moves.isna().sum().sum() / moves.size * 100)

7389
1237
16.74110163756936


## COMPLETEZZA ABILITIES

In [11]:
ability = pd.read_csv('Dataset integrati 2.0/df_abilities.csv')
ability

Name  Generation  \
0    Adaptability           4   
1        Aerilate           6   
2       Aftermath           4   
3        Air Lock           3   
4        Analytic           5   
..            ...         ...   
262   White Smoke           3   
263      Wimp Out           7   
264  Wonder Guard           3   
265   Wonder Skin           5   
266      Zen Mode           5   

                                          Description  
0                   Powers up moves of the same type.  
1     Turns Normal-type moves into Flying-type moves.  
2     Damages the attacker landing the finishing hit.  
3                  Eliminates the effects of weather.  
4      Boosts move power when the Pokémon moves last.  
..                                                ...  
262   Prevents other Pokémon from lowering its stats.  
263            Switches out when HP drops below half.  
264               Only supereffective moves will hit.  
265  Makes status-changing moves more likely to miss.  
266            Changes form when HP drops below half.  

[267 rows x 3 columns]

In [12]:
ability_column_nan = pd.DataFrame(ability.isna().sum(), columns=['Total NaN'])
ability_column_nan

Total NaN
Name                 0
Generation           0
Description          1

In [13]:
#Calcolo la percentuale di NaN per singola colonna, considerando che abbiamo 267 righe
ability_column_nan['% NaN'] = (ability_column_nan['Total NaN']/267)*100
ability_column_nan = ability_column_nan.round(decimals=2)
ability_column_nan

Total NaN  % NaN
Name                 0   0.00
Generation           0   0.00
Description          1   0.37

Abbiamo solamente una abilità per cui manca la descrizione.

In [54]:
## Completezza intero dataset
print(ability.size)
print(ability.isna().sum().sum())
print(ability.isna().sum().sum() / ability.size * 100)

801
1
0.12484394506866417


## COMPLETEZZA NATURES

In [14]:
nature = pd.read_csv('Dataset integrati 2.0/df_natures.csv')
nature

Nature Increases Decreases Likes_berrie Dislikes_berrie
0   Adamant    Attack   Sp. Atk        Spicy             Dry
1   Bashful   Sp. Atk   Sp. Atk          Dry             Dry
2      Bold   Defense    Attack         Sour           Spicy
3     Brave    Attack     Speed        Spicy           Sweet
4      Calm   Sp. Def    Attack       Bitter           Spicy
5   Careful   Sp. Def   Sp. Atk       Bitter             Dry
6    Docile   Defense   Defense         Sour            Sour
7    Gentle   Sp. Def   Defense       Bitter            Sour
8     Hardy    Attack    Attack        Spicy           Spicy
9     Hasty     Speed   Defense        Sweet            Sour
10   Impish   Defense   Sp. Atk         Sour             Dry
11    Jolly     Speed   Sp. Atk        Sweet             Dry
12      Lax   Defense   Sp. Def         Sour          Bitter
13   Lonely    Attack   Defense        Spicy            Sour
14     Mild   Sp. Atk   Defense          Dry            Sour
15   Modest   Sp. Atk    Attack          Dry           Spicy
16    Naive     Speed   Sp. Def        Sweet          Bitter
17  Naughty    Attack   Sp. Def        Spicy          Bitter
18    Quiet   Sp. Atk     Speed          Dry           Sweet
19   Quirky   Sp. Def   Sp. Def       Bitter          Bitter
20     Rash   Sp. Atk   Sp. Def          Dry          Bitter
21  Relaxed   Defense     Speed         Sour           Sweet
22    Sassy   Sp. Def     Speed       Bitter           Sweet
23  Serious     Speed     Speed        Sweet           Sweet
24    Timid     Speed    Attack        Sweet           Spicy

In [15]:
natures_column_nan = pd.DataFrame(nature.isna().sum(), columns=['Total NaN'])
natures_column_nan

Total NaN
Nature                   0
Increases                0
Decreases                0
Likes_berrie             0
Dislikes_berrie          0

Non ci sono NaN nel dataset delle nature.

## COMPLETEZZA TYPES

In [16]:
types = pd.read_csv('Dataset integrati 2.0/df_types.csv')
types

Name  Generation
0     normal           1
1   fighting           1
2     flying           1
3     poison           1
4     ground           1
5       rock           1
6        bug           1
7      ghost           1
8      steel           2
9       fire           1
10     water           1
11     grass           1
12  electric           1
13   psychic           1
14       ice           1
15    dragon           1
16      dark           2
17     fairy           6
18   unknown           2
19    shadow           3

In [17]:
types_column_nan = pd.DataFrame(types.isna().sum(), columns=['Total NaN'])
types_column_nan

Total NaN
Name                0
Generation          0

Non ci sono NaN nel dataset dei tipi.